core-services/src/main/java/org/zalando/nakadi/service/publishing/EventsProcessor.java

```diff
@@ -10,84 +10,183 @@
+        private final JSONArray data;
+        private int size = 0;
+
+        private BatchedRequest(final String eventType, final long finishCollectionAt) {
+            this.eventType = eventType;
+            this.finishCollectionAt = finishCollectionAt;
+            this.data = new JSONArray();
+        }
+
+        public int add(final JSONObject obj) {
+            this.data.put(obj);
+            return ++size;
+        }
 
-            final long finishAt = System.currentTimeMillis() + batchCollectionTimeout;
-            int eventsCount = 0;
-            JSONArray jsonArray = null;
-            while (eventsCount != batchSize && System.currentTimeMillis() < finishAt) {
+        @Override
+        public String toString() {
+            return "Batch{et=" + eventType + ",size:" + size + "}";
+        }
+    }
+
+    private void scheduleSendBatchedRequest(final BatchedRequest req) {
+        final Runnable r = new Runnable() {
+            @Override
+            public void run() {
                 try {
-                    final JSONObject event = events.poll(pollTimeout, TimeUnit.MILLISECONDS);
-                    if (event != null) {
-                        if (jsonArray == null) {
-                            jsonArray = new JSONArray();
+                    eventPublisher.processInternal(req.data.toString(), req.eventType, false, null, false);
+                } catch (final RuntimeException ex) {
+                    LOG.info("Failed to send single batch for unknown reason", ex);
+                }
+            }
+
+            @Override
+            public String toString() {
+                return "Batch to " + req.eventType + " of size " + req.data.length();
+            }
+        };
+        executorService.submit(r);
+    }
+
+    private void dispatch() {
+        final Map<String, BatchedRequest> batchesBeingAssembled = new HashMap<>();
+        try {
+            long nextTimeCheck = System.currentTimeMillis() + batchCollectionTimeout;
+            while (true) {
+                long currentTime = System.currentTimeMillis();
+                final EventToPublish data = eventsQueue.poll(
+                        Math.max(nextTimeCheck - currentTime, 1), TimeUnit.MILLISECONDS);
+                currentTime = System.currentTimeMillis();
+                boolean batchWasSent = false;
+                if (data != null) {
+                    BatchedRequest batch = batchesBeingAssembled.get(data.eventType);
+                    if (null == batch) {
+                        batch = new BatchedRequest(data.eventType, currentTime + batchCollectionTimeout);
+                        batchesBeingAssembled.put(data.eventType, batch);
+                    }
+                    if (batch.add(data.object) >= maxBatchSize) {
+                        scheduleSendBatchedRequest(batch);
+                        batchesBeingAssembled.remove(batch.eventType);
+                        currentTime = System.currentTimeMillis();
```

Comment:
since it is current time, probably, should be the last operation in this if clause